### ISSUE TITLE
[Feature]: Better Tailwind Customisation

### ISSUE THREAD

**Issue by thombruce** [thombruce/tnt-2024#65]
### Feature request

Related to #15 

I want to quote this from that thread:

> There are a couple of approaches we might try:
>
> 1. Using a different "preset": https://tailwindcss.com/docs/configuration#presets
> 2. Using multiple configurations: https://tailwindcss.com/docs/configuration#using-multiple-configurations

These are still both worth looking into, though either one probably does need some careful consideration with respect to handling via a module.

The ideal solution would allow the user of TNT to have their own `tailwind.config[.js|.ts]` file in the root of their project, and have this serve as the basis for their style overrides. _This would have in fact been the default behaviour, had I not overwritten it to favour the TNT layer's own tailwind config._ This (their own config) should have priority given to its values when merging.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by thombruce** [comment]
I've looked at the options, 1 and 2 above, and I think I can loosely conclude that number 2 is a non-starter. This approach allows the user to have **different** configs per entry-point. That's no good. We're looking for a solution which allows a user's config to be merged with and override TNT's defaults.

Option 1 is slightly more promising and worth a look. Instead of TNT having a config file, it would instead have a preset. It's pretty much the same thing, except... it has lesser priority than the config?

Now, maybe it means the end user needs to include our preset in their config file but... that should be fine!

It may also be possible to merge it as a preset via the module approach... though I was having trouble with housing the module on tnt/core previously, particularly around plugins. Could be worth a look to see if preset does work this way.

**Comment by thombruce** [comment]
Presets (option 1) is also proving to be... difficult.

I can get it to work! It works! But it works with this issue:

> **WARN** A hook has injected a non-serializable value in config["presets"], so the Tailwind Config cannot be serialized. Falling back to providing the loaded configuration inlined directly to PostCSS loader.. Please consider using a configuration file/template instead (specifying in configPath of the module options) to enable additional support for IntelliSense and HMR.

But... it does work. So... here's the module code that's functionally working (despite that warning) in case I circle back to this:

```ts
import { defineNuxtModule } from '@nuxt/kit'

const colors = require('tailwindcss/colors')

export default defineNuxtModule({
  setup (options, nuxt) {
    nuxt.hook('tailwindcss:config', function (tailwindConfig) {
      tailwindConfig.presets =
        tailwindConfig.presets?.concat([require('../tailwind')]) ||
        [require('../tailwind')]
    })
  }
})
```

What I was thinking of attempting next was... something I actually already know gives the same warning. In fact, I've just experimented with removing the particular configuration responsible from the preset used with the above module code and...

- Without `plugins: [...]` the warning is gone (but some site CSS is otherwise broken)

The main drawback of the warning shown is... no more HMR (Hot Module Reloading) which is particularly useful for styles, as you want to be able to tweak and see this update as quickly as possible.

So it's a no-go.

There's no way around it, the presets approach cannot work unless I inform users that they **must** include those plugins in their own config, and then...  then I'm sacrificing the biggest convenience this alternative had in the first place.

It's just not a workable solution.

**Comment by thombruce** [comment]
The best approach... perhaps only reasonable approach... other than the module approach which is where we're at right now is...

We provide a `tailwind.preset.js` in TNT/core but we don't automatically try to merge it. Instead, we encourage the user to include in their own tailwind config:

```js
export default {
  presets: [
    require('@thombruce/tnt/tailwind.preset'),
  ],
  // ...
}
```

This loads the preset, which their own config can then add to.

No warnings about non-serializable objects this way. It's a lot more error-resistant than the modules approach. But... right now it means that TNT won't work until this initial setup step is taken.

---

Here's what I'm thinking...

Can TNT check to see if there's a `tailwind.config` in the user's project AND **if not**... use TNT's `tailwind.preset` as though it were a config file (because it is - they are same schema)?

That might be the best approach. Perhaps we can do this directly from `nuxt.config`...

**Comment by thombruce** [comment]
We can indeed arrange to have that work directly from `nuxt.config`. Here's the solution:

```ts
  tailwindcss: {
    configPath: fs.existsSync('./tailwind.config.js') ? './tailwind.config.js'
      : fs.existsSync('./tailwind.config.ts') ? './tailwind.config.ts'
      : join(currentDir, './tailwind.preset.js'),
    cssPath: join(currentDir, './assets/css/tnt.css'),
  },
```

It's checking the file system at the root of the end user's project for a `tailwind.config.js` or `tailwind.config.ts` file and using either of these if they're present. If not, it uses the folder path for the TNT Nuxt config file (`const currentDir = dirname(fileURLToPath(import.meta.url))`) and joins that with the `tailwind.preset.js` to use that instead.

The end user **DOES** still need to include our preset, but **ONLY** if they intend to extend/restyle Tailwind from our defaults. If they simply don't have a Tailwind config file, ours will be used.

It's shy of perfect, but just barely! It's error prone, sure, but significantly less error prone than the module solution I was using previously.

**Comment by thombruce** [comment]
Fixed: 2c45373276c2ea8b9d1c7d7efb77ba475e8c2cce
